### ISSUE TITLE
Reduce memory requirements for dereplication

### ISSUE THREAD

**Issue by torognes** [torognes/vsearch#329]
When running `derep_fulllength` VSEARCH initially reads all the input sequences into memory before dereplicating them. By reading one sequence at a time and storing only the unique sequences there should be great opportunities to reduce the memory footprint.

**Comment by torognes** [comment]
Version 2.8.2 will now read one sequence at a time and only store the distinct sequences, potentially saving memory when there are many duplicates.

**Comment by mys721tx** [comment]
I used massif from Valgrind to profile the memory usage. The dataset is the MiSeq_SOP dataset. The patch seems to increase the peak heap allocation instead of reducing it.

|commit| peak heap allocation/B | peak snapshot |
|---|---|---|
|7c83b49| 36,915,176| 31 |
| 15f715a| 127,889,424|46|

The profiles can be found [in this gist](https://gist.github.com/mys721tx/b156e9ad5705b91783547363a43269ba).

**Comment by torognes** [comment]
@mys721tx , it is correct that the new approach may use some more memory with small datasets or with datasets that hardly contain duplicates. However, with large datasets with lots of duplicates, I believe the new approach should save much memory, and it is in those situations it really matters. Could you test it again on a large and more realistic dataset?

**Comment by frederic-mahe** [comment]
@mys721tx thanks for profiling the memory usage. For replicability, could you please post the command line you used to produce the profile?

**Comment by mys721tx** [comment]
Valgrind version is valgrind-3.14.0.GIT. The command I have used is:

```ShellSession
valgrind --tool=massif --stacks=yes ./vsearch --derep_fulllength all.fasta --sizein --sizeout --output all.derep.test.fasta
```

**Comment by mys721tx** [comment]
@torognes I don't think it is caused by the low duplication level in the dataset. `all.fasta` is 6.8M. The allocation on the heap before the patch is around 30M. It would be enough to store the whole input if every sequence in it is unique.

**Comment by torognes** [comment]
Dereplication will now allocate memory for a million sequences in one of the hash tables and grow it if necessary. Therefore, you may not see the effect if the dataset is smaller.

**Comment by colinbrislawn** [comment]
You beat me to the benchmark! 

I think the typical use case would be dereplicating raw or paired end reads. Here's the raw forward read from ITS amplicon on Illumina MiSeq. 


|     | version |   nucleotides  | sequences | gz size | peak heap | result |
|---|--------|----------------|----------|----------|-----------|------|
| full R1 read | 2.8.2 | 4,775,787,793 | 19,027,043 | 3.8 GB | 4.68 GB  | success! |
| full R1 read | 2.8.1 | 4,775,787,793 | 19,027,043 | 3.8 GB | 9.68 GB | error | 


The second run ended with the following error when at 86% finished:


```
vsearch v2.8.1_linux_x86_64, 62.9GB RAM, 24 cores
vsearch --derep_fulllength Undetermined_S0_L001_R1_001.fastq.gz --sizein --sizeout --output R1_older.derep.fasta --log R1_older.derep.log
Started  Fri Aug 24 11:24:19 2018


Fatal error: Unable to reallocate enough memory.
```

**Comment by torognes** [comment]
I have made changes to reduce memory requirements further. See issue #334. Released in version 2.8.4.

**Comment by frederic-mahe** [comment]
`vsearch`'s memory-footprint when dereplicating is now much closer to the size of the file **after** dereplication, rather than being closer to the size of the file **before** dereplication. Awesome!
